File: 9.speaker_play_demo/python/speaker_play_demo/speaker_play_demo_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Speaker Play Demo Node (Python version)

通过 /intelligent_interaction/tts/play 服务（interaction_msgs/srv/TtsService）
实现音频播放控制，支持文本/文件/URL三种播报方式。
"""

import os
import threading

import rclpy
from rclpy.node import Node
from interaction_msgs.srv import TtsService


class SpeakerPlayDemoNode(Node):
    """ROS2 Node for speaker play demo via TtsService"""

    def __init__(self):
        super().__init__('speaker_play_demo_node')

        self.declare_parameter('default_text', '你好，我是天工机器人。')
        self.declare_parameter('default_audio_path', '/tmp/test.wav')
        self.declare_parameter('auto_play_on_start', False)

        self.default_text = self.get_parameter('default_text').get_parameter_value().string_value
        self.default_audio_path = self.get_parameter('default_audio_path').get_parameter_value().string_value
        self.auto_play_on_start = self.get_parameter('auto_play_on_start').get_parameter_value().bool_value

        self.tts_client = self.create_client(
            TtsService, '/intelligent_interaction/tts/play')

        self.get_logger().info('Speaker Play Demo Node initialized')
        self.get_logger().info(f'  Default text: {self.default_text}')
        self.get_logger().info(f'  Default audio path: {self.default_audio_path}')

        self.get_logger().info('Waiting for TTS service...')
        if self.tts_client.wait_for_service(timeout_sec=5.0):
            self.get_logger().info('TTS service is ready')
        else:
            self.get_logger().warn('TTS service not available, will retry when needed')

        if self.auto_play_on_start:
            def delayed_play():
                import time
                time.sleep(1)
                self.play_text(self.default_text)
            threading.Thread(target=delayed_play, daemon=True).start()

    def _call_tts(self, text: str, res_type: str, cmd: str):
        """Call TTS service"""
        if not self.tts_client.service_is_ready():
            self.get_logger().error('TTS service is not ready')
            return

        request = TtsService.Request()
        request.text = text
        request.type = res_type
        request.cmd = cmd

        self.get_logger().info(
            f'TTS call: text="{text[:50]}", type={res_type}, cmd={cmd}')

        future = self.tts_client.call_async(request)
        future.add_done_callback(self._tts_response_callback)

    def _tts_response_callback(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'TTS success. Status: {response.status}')
            else:
                self.get_logger().error(f'TTS failed. Status: {response.status}')
        except Exception as e:
            self.get_logger().error(f'TTS service call failed: {e}')

    def play_text(self, text: str):
        """Play text via TTS"""
        self._call_tts(text, 'text', 'append')

    def play_file(self, path: str):
        """Play audio file"""
        if not os.path.exists(path):
            self.get_logger().error(f'Audio file does not exist: {path}')
            return
        self._call_tts(path, 'file', 'append')

    def play_url(self, url: str):
        """Play audio from URL"""
        self._call_tts(url, 'url', 'append')

    def stop(self):
        """Stop current playback"""
        self._call_tts('', 'text', 'stop')

    def query_status(self):
        """Query playback status"""
        self._call_tts('', 'text', 'query')


def print_help():
    print('\n========================================')
    print('Speaker Play Demo - Commands:')
    print('  text <content>  - Play text via TTS')
    print('  play <path>     - Play audio file')
    print('  url <url>       - Play audio from URL')
    print('  stop            - Stop playback')
    print('  status          - Query playback status')
    print('  help            - Show this help')
    print('  quit            - Exit program')
    print('========================================')


def main(args=None):
    rclpy.init(args=args)
    node = SpeakerPlayDemoNode()

    spin_thread = threading.Thread(target=rclpy.spin, args=(node,), daemon=True)
    spin_thread.start()

    try:
        print_help()

        while rclpy.ok():
            try:
                line = input()
                if not line:
                    continue

                parts = line.split(maxsplit=1)
                cmd = parts[0]

                if cmd == 'text':
                    if len(parts) > 1:
                        node.play_text(parts[1])
                    else:
                        node.play_text(node.default_text)
                elif cmd == 'play':
                    if len(parts) > 1:
                        node.play_file(parts[1])
                    else:
                        node.play_file(node.default_audio_path)
                elif cmd == 'url':
                    if len(parts) > 1:
                        node.play_url(parts[1])
                    else:
                        print('Usage: url <url>')
                elif cmd == 'stop':
                    node.stop()
                elif cmd == 'status':
                    node.query_status()
                elif cmd == 'help':
                    print_help()
                elif cmd in ('quit', 'exit'):
                    break
                else:
                    print(f'Unknown command: {cmd}')
                    print_help()
            except EOFError:
                break
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()
        spin_thread.join(timeout=1.0)


if __name__ == '__main__':
    main()

File: speaker_play_demo/cpp/speaker_play_demo/src/speaker_play_demo_node.cpp
/**
 * @file speaker_play_demo_node.cpp
 * @brief 通过 /intelligent_interaction/tts/play 服务实现音频播放控制
 *
 * 支持文本/文件/URL三种播报方式和停止/查询控制
 */

#include <memory>
#include <string>
#include <fstream>
#include <iostream>
#include <sstream>
#include <thread>
#include <chrono>

#include "rclcpp/rclcpp.hpp"
#include "interaction_msgs/srv/tts_service.hpp"

class SpeakerPlayDemoNode : public rclcpp::Node {
 public:
  SpeakerPlayDemoNode() : Node("speaker_play_demo_node") {
    this->declare_parameter("default_text", "你好，我是天工机器人。");
    this->declare_parameter("default_audio_path", "/tmp/test.wav");
    this->declare_parameter("auto_play_on_start", false);

    default_text_ = this->get_parameter("default_text").as_string();
    default_audio_path_ = this->get_parameter("default_audio_path").as_string();
    auto_play_on_start_ = this->get_parameter("auto_play_on_start").as_bool();

    tts_client_ = this->create_client<interaction_msgs::srv::TtsService>(
        "/intelligent_interaction/tts/play");

    RCLCPP_INFO(this->get_logger(), "Speaker Play Demo Node initialized");
    RCLCPP_INFO(this->get_logger(), "  Default text: %s", default_text_.c_str());
    RCLCPP_INFO(this->get_logger(), "  Default audio path: %s", default_audio_path_.c_str());

    RCLCPP_INFO(this->get_logger(), "Waiting for TTS service...");
    if (!tts_client_->wait_for_service(std::chrono::seconds(5))) {
      RCLCPP_WARN(this->get_logger(), "TTS service not available, will retry when needed");
    } else {
      RCLCPP_INFO(this->get_logger(), "TTS service is ready");
    }

    if (auto_play_on_start_) {
      std::thread([this]() {
        std::this_thread::sleep_for(std::chrono::seconds(1));
        play_text(default_text_);
      }).detach();
    }
  }

  void call_tts(const std::string& text, const std::string& res_type, const std::string& cmd) {
    if (!tts_client_->service_is_ready()) {
      RCLCPP_ERROR(this->get_logger(), "TTS service is not ready");
      return;
    }

    auto request = std::make_shared<interaction_msgs::srv::TtsService::Request>();
    request->text = text;
    request->type = res_type;
    request->cmd = cmd;

    RCLCPP_INFO(this->get_logger(), "TTS call: text=\"%.50s\", type=%s, cmd=%s",
                text.c_str(), res_type.c_str(), cmd.c_str());

    tts_client_->async_send_request(
        request,
        [this](rclcpp::Client<interaction_msgs::srv::TtsService>::SharedFuture future) {
          auto response = future.get();
          if (response->success) {
            RCLCPP_INFO(this->get_logger(), "TTS success. Status: %s",
                        response->status.c_str());
          } else {
            RCLCPP_ERROR(this->get_logger(), "TTS failed. Status: %s",
                         response->status.c_str());
          }
        });
  }

  void play_text(const std::string& text) {
    call_tts(text, "text", "append");
  }

  void play_file(const std::string& path) {
    std::ifstream f(path);
    if (!f.good()) {
      RCLCPP_ERROR(this->get_logger(), "Audio file does not exist: %s", path.c_str());
      return;
    }
    call_tts(path, "file", "append");
  }

  void play_url(const std::string& url) {
    call_tts(url, "url", "append");
  }

  void stop() {
    call_tts("", "text", "stop");
  }

  void query_status() {
    call_tts("", "text", "query");
  }

 private:
  rclcpp::Client<interaction_msgs::srv::TtsService>::SharedPtr tts_client_;
  std::string default_text_;
  std::string default_audio_path_;
  bool auto_play_on_start_;
};

void print_help() {
  std::cout << "\n========================================" << std::endl;
  std::cout << "Speaker Play Demo - Commands:" << std::endl;
  std::cout << "  text <content>  - Play text via TTS" << std::endl;
  std::cout << "  play <path>     - Play audio file" << std::endl;
  std::cout << "  url <url>       - Play audio from URL" << std::endl;
  std::cout << "  stop            - Stop playback" << std::endl;
  std::cout << "  status          - Query playback status" << std::endl;
  std::cout << "  help            - Show this help" << std::endl;
  std::cout << "  quit            - Exit program" << std::endl;
  std::cout << "========================================" << std::endl;
}

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SpeakerPlayDemoNode>();

  std::thread spin_thread([&node]() {
    rclcpp::spin(node);
  });

  if (argc > 1) {
    std::string cmd = argv[1];
    if (cmd == "text" && argc > 2) {
      node->play_text(argv[2]);
    } else if (cmd == "play" && argc > 2) {
      node->play_file(argv[2]);
    } else if (cmd == "url" && argc > 2) {
      node->play_url(argv[2]);
    } else if (cmd == "stop") {
      node->stop();
    } else if (cmd == "status") {
      node->query_status();
    }
    std::this_thread::sleep_for(std::chrono::seconds(1));
  } else {
    print_help();

    std::string line;
    while (rclcpp::ok() && std::getline(std::cin, line)) {
      if (line.empty()) continue;

      std::istringstream iss(line);
      std::string cmd;
      iss >> cmd;

      if (cmd == "text") {
        std::string rest;
        std::getline(iss >> std::ws, rest);
        if (!rest.empty()) {
          node->play_text(rest);
        } else {
          std::cout << "Usage: text <content>" << std::endl;
        }
      } else if (cmd == "play") {
        std::string path;
        if (iss >> path) {
          node->play_file(path);
        } else {
          std::cout << "Usage: play <path>" << std::endl;
        }
      } else if (cmd == "url") {
        std::string url;
        if (iss >> url) {
          node->play_url(url);
        } else {
          std::cout << "Usage: url <url>" << std::endl;
        }
      } else if (cmd == "stop") {
        node->stop();
      } else if (cmd == "status") {
        node->query_status();
      } else if (cmd == "help") {
        print_help();
      } else if (cmd == "quit" || cmd == "exit") {
        break;
      } else {
        std::cout << "Unknown command: " << cmd << std::endl;
        print_help();
      }
    }
  }

  rclcpp::shutdown();
  if (spin_thread.joinable()) spin_thread.join();
  return 0;
}
